### ISSUE TITLE
Challanges We Encountered as a Group

### ISSUE THREAD

**Issue by SKutayKilic** [bounswe/bounswe2019group7#101]
Challanges We Encountered as a Group of milestone will be prepared.
Deadline: 11.05.2019, 23:59

**Comment by safakozdek** [comment]
As a suggestion, for the problem of creating a new dashboard to manage email lists we conducted a research and agreed on using Mail Chimp because it offers great functionality and has an API that is very easy to use.

**Comment by SKutayKilic** [comment]
**Challanges We Encountered as a Group**

At the start of the api project we had to choose technologies to use in our api. However, it was not easy for us because we did not know the advantages and disadvantages of the frameworks which we had to choose. After a rigorous research on the choices, we choose Django because of it’s object relational mapping system, ease of migration, elegant routing system, and powerful serialization/deserialization libraries. 

After we have chosen the framework to use, we had to find the functionalities to implement. In order to deciding the functionalities we made a brainstorming considering the restrictions such as the functionalities must be related to the Traders Project, some functionalities must use external apis, some must do crud operations on the database, and they should use different http methods(POST, GET, PUT,DELETE, etc) which are proper for the requirements. Also, we needed to share those functionalities as fair as possible so we decided to who will implement which functionality in our second meeting about the api project.

When we started to the implementation part, most of us did not know very well about the RESTful Apis, MVT architecture, and some important libraries of Django so we taught those important knowledge to each other. Furthermore, the bugs related to routing system, migrations, and response formats were frequently occurred so we helped each other to fix those bugs. Finally, we had a working api that is ready to use.

The api part of the project was not sufficient because we had to implement also a user interface. However, since learning to design a frontend application is long commitment we have chosen a subgroup consisting of two people who have experience on frontend development. Our subgroup, started to implement the frontend application using Reactjs. They encountered some challenges because of the asynchronous nature of the javascript. For example, they encountered some binding errors because bindings had to wait for the response from our api but they overcome that situation using the built-in lifecycle methods in Reactjs. At the end, we had both backend and frontend application which are working properly together.
